### PR TITLE
Associations

### PR DESCRIPTION
**Pull request by AlexRS90**
## For this milestone I:
✅  Created a class <code>Classroom</code> 🧑‍🏫 
✅  Created the <code>has-many/belongs-to</code> relationship between <code>Classroom</code> and <code>Student</code> ◀️ ▶️ 
✅  Created a class <code>Book</code> 📚 
✅  Create a class <code>Rental</code> 🧾 
✅  Create the <code>many-to-many</code> relationship between <code>Person</code> and <code>Book</code> using the intermediate class <code>Rental</code> ⏫  ⏬

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -25,6 +25,14 @@ Override <code>can_use_services?</code> method thanks to polymorphism
...
 ✅ Added a method called <code>validate_name</code> to update <code>@name</code> using <code>Corrector</code> instance 🆙 <br>
 
+## Milestone 3:
+
+✅  Created a class <code>Classroom</code> 🧑‍🏫  <br>
+✅  Created the <code>has-many/belongs-to</code> relationship between <code>Classroom</code> and <code>Student</code> ◀️ ▶️ <br>
+✅  Created a class <code>Book</code> 📚  <br>
+✅  Create a class <code>Rental</code> 🧾 <br>
+✅  Create the <code>many-to-many</code> relationship between <code>Person</code> and <code>Book</code> using the intermediate class <code>Rental</code> ⏫  ⏬ <br>
+
```
Comment (julie-ify) [inline README.md:35]: - [x] Your README is so descriptive 👍

File: book.rb [hunk 2]
```diff
@@ -0,0 +1,10 @@
+class Book
+  attr_accessor :title, :author
+  attr_reader :rental
```
Comment (julie-ify) [inline book.rb]: - [x] Let's add `:rental` to attr_accessor as users can write and read the rental. See suggestion below:

```suggestion
  attr_accessor :title, :author, :rental
```

File: book.rb [hunk 3]
```diff
@@ -0,0 +1,10 @@
...
+
+  def initialize(title, author)
+    @title = title
+    @author = author
+    @rental = []
+  end
+end
```
Comment (julie-ify) [inline book.rb]: - [x] Let's create a method for adding rentals. The book itself should be sent as a parameter to create the new rental achieving the 'has-many' association. Check out this [article](https://github.com/microverseinc/curriculum-ruby/blob/main/oop/articles/oop_relationships_by_examples.md#many-to-many-relationship) as a reference.

File: person.rb [hunk 4]
```diff
@@ -2,14 +2,15 @@
 
 class Person
   attr_accessor :name, :age
-  attr_reader :id
+  attr_reader :id, :rental
```
Comment (julie-ify) [inline person.rb]: - [ ] Let's add `:rental` to attr_accessor as users can write and read the rental.

File: person.rb [hunk 5]
```diff
@@ -2,14 +2,15 @@
...
 
   def initialize(age, name = 'Unknown', parent_permission: true)
     @id = Random.rand(1..500)
     @name = name
     @age = age
     @parent_permission = parent_permission
     @check_name = Corrector.new
+    @rental = []
   end
```
Comment (julie-ify) [inline person.rb]: - [ ] Let's create a method for adding rental. The person itself should be sent as a parameter to create the new rental achieving the 'has-many' association. Check out this [lesson](https://github.com/microverseinc/curriculum-ruby/blob/main/oop/articles/oop_relationships_by_examples.md#many-to-many-relationship) as reference.

File: rental.rb [hunk 6]
```diff
@@ -0,0 +1,14 @@
+class Rental
+  attr_accessor :date
+  attr_reader :person, :book
```
Comment (julie-ify) [inline rental.rb]: - [x]  Let's add `:book` and `:person` to attr_accessor as users can write and read them. See suggestion below:

```suggestion
  attr_accessor :date, :person, :book
```

**Review by julie-ify — CHANGES_REQUESTED** [review]
## Hi @AlexRS90,

Good job so far!

## Highlight!
- Good feature branch name ✔️ 
- No linter errors ✔️ 
- Descriptive README ✔️ 
- Descriptive pull request ✔️ 
- Descriptive commit messages ✔️ 
- Good code organization ✔️

Hello ✋, you have done well reaching this point of the milestone 🤝 . But there are some issues that you still need to work on to go to the next project but you are almost there!

# Status: Required Changes ♻️⛔️
_Check the comments under the review._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: README.md (hunk 1 above)
Comment (AlexRS90) [reply README.md:35]: Thank you @julie-ify 😄

File: book.rb (hunk 2 above)
Comment (AlexRS90) [reply book.rb]: Thanks for the suggestion I'll implement it!

File: book.rb (hunk 3 above)
Comment (AlexRS90) [reply book.rb]: Oh, it is true! I missed that method, thanks!

File: person.rb (hunk 4 above)
Comment (AlexRS90) [reply person.rb]: Thanks for the suggestion I'll implement it!

File: person.rb (hunk 5 above)
Comment (AlexRS90) [reply person.rb]: Oh, it is true! I missed that method, thanks!

File: book.rb [hunk 7]
```diff
@@ -0,0 +1,14 @@
...
+
+  def initialize(title, author)
+    @title = title
+    @author = author
+    @rental = []
+  end
+
+  def add_rental(book)
+    @rental.push(book)
+    rental.book = self
```
Comment (SajjadAhmad14) [inline book.rb]: Please create a new rental in `#add_rental` of the `Book` class and pass the book as a `self`, person, and date as parameters. You may see what parameters are needed by Rental's class constructor to create an object in order to pass the right params.

File: person.rb [hunk 8]
```diff
@@ -20,6 +21,11 @@ def can_use_services?
     (of_age? || @parent_permission)
   end
 
+  def add_rentals(rental)
+    rental << rental
+    rental.person = self
```
Comment (SajjadAhmad14) [inline person.rb]: Please create a new rental in `#add_rental` of `Person` class and pass person as a `self`, book, and date as parameters. You may see what parameters are needed by Rental's class constructor to create an object in order to pass the right params.

**Review by SajjadAhmad14 — CHANGES_REQUESTED** [review]
## Required Changes ♻️
> "Change is inevitable. Change is constant. -Benjamin Disraeli"
## Hi Alex,
### Great work on making the changes requested by a previous reviewer 👏 
### You've done well implementing some of the requested changes, but there are still some which aren't been addressed yet.

## To Improve 🙏

- [ ] Please check inline comments under the review

## Optional suggestions
_Comments with the [OPTIONAL] prefix are not a must to implement however, It's strongly recommended to work on them as they can make your code better._

## Happy coding!👏👏👏

#### Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
#### Please remember to tag me in your question so I can receive the notification.
**_Note: Please do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s).

File: book.rb (hunk 7 above)
Comment (AlexRS90) [reply book.rb]: Hi @SajjadAhmad14 thanks for reviewing my project, but I'm not really understanding what I have to do, how can I pass date as a parameter from <code>book</code> class?

Comment (SajjadAhmad14) [reply book.rb]: Hi, Thank you for understanding and making changes!

**Review by SajjadAhmad14 — APPROVED** [review]
# Status: Approved :heavy_check_mark: :heavy_check_mark: :heavy_check_mark:

### Hi Alex, 
## Congratulations 🎉 Thank you for implementing all the requirements 🙏
### Please go ahead and merge the branch!

## Happy Coding 👏👏👏
#### Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s).

**Comment by AlexRS90** [comment]
Thank you very much! 🤓

**Pull request merged**